src/data/set/finite.lean

```diff
@@ -20,529 +47,733 @@ variables {α : Type u} {β : Type v} {ι : Sort w} {γ : Type x}
+structure. -/
+def fintype_bUnion [decidable_eq α] {ι : Type*} (s : set ι) [fintype s]
+  (t : ι → set α) (H : ∀ i ∈ s, fintype (t i)) : fintype (⋃(x ∈ s), t x) :=
+fintype.of_finset
+(s.to_finset.attach.bUnion
+  (λ x, by { haveI := H x (by simpa using x.property), exact (t x).to_finset })) $ by simp
 
-local attribute [instance] decidable_mem_of_fintype
+instance fintype_bUnion' [decidable_eq α] {ι : Type*} (s : set ι) [fintype s]
+  (t : ι → set α) [∀ i, fintype (t i)] : fintype (⋃(x ∈ s), t x) :=
+fintype.of_finset (s.to_finset.bUnion (λ x, (t x).to_finset)) $ by simp
 
-instance fintype_insert [decidable_eq α] (a : α) (s : set α) [fintype s] :
-  fintype (insert a s : set α) :=
-if h : a ∈ s then by rwa [insert_eq, union_eq_self_of_subset_left (singleton_subset_iff.2 h)]
-else fintype_insert' _ h
+/-- If `s : set α` is a set with `fintype` instance and `f : α → set β` is a function such that
+each `f a`, `a ∈ s`, has a `fintype` structure, then `s >>= f` has a `fintype` structure. -/
+def fintype_bind {α β} [decidable_eq β] (s : set α) [fintype s]
+  (f : α → set β) (H : ∀ a ∈ s, fintype (f a)) : fintype (s >>= f) :=
+set.fintype_bUnion s f H
 
-end
+instance fintype_bind' {α β} [decidable_eq β] (s : set α) [fintype s]
+  (f : α → set β) [H : ∀ a, fintype (f a)] : fintype (s >>= f) :=
+set.fintype_bUnion' s f
 
-@[simp] theorem finite.insert (a : α) {s : set α} : finite s → finite (insert a s)
-| ⟨h⟩ := ⟨@set.fintype_insert _ (classical.dec_eq α) _ _ h⟩
+instance fintype_empty : fintype (∅ : set α) := fintype.of_finset ∅ $ by simp
 
-lemma to_finset_insert [decidable_eq α] {a : α} {s : set α} (hs : finite s) :
-  (hs.insert a).to_finset = insert a hs.to_finset :=
-finset.ext $ by simp
+instance fintype_singleton (a : α) : fintype ({a} : set α) := fintype.of_finset {a} $ by simp
 
-@[simp] lemma insert_to_finset [decidable_eq α] {a : α} {s : set α} [fintype s] :
-  (insert a s).to_finset = insert a s.to_finset :=
-by simp [finset.ext_iff, mem_insert_iff]
+instance fintype_pure : ∀ a : α, fintype (pure a : set α) :=
+set.fintype_singleton
 
-@[elab_as_eliminator]
-theorem finite.induction_on {C : set α → Prop} {s : set α} (h : finite s)
-  (H0 : C ∅) (H1 : ∀ {a s}, a ∉ s → finite s → C s → C (insert a s)) : C s :=
-let ⟨t⟩ := h in by exactI
-match s.to_finset, @mem_to_finset _ s _ with
-| ⟨l, nd⟩, al := begin
-    change ∀ a, a ∈ l ↔ a ∈ s at al,
-    clear _let_match _match t h, revert s nd al,
-    refine multiset.induction_on l _ (λ a l IH, _); intros s nd al,
-    { rw show s = ∅, from eq_empty_iff_forall_not_mem.2 (by simpa using al),
-      exact H0 },
-    { rw ← show insert a {x | x ∈ l} = s, from set.ext (by simpa using al),
-      cases multiset.nodup_cons.1 nd with m nd',
-      refine H1 _ ⟨finset.subtype.fintype ⟨l, nd'⟩⟩ (IH nd' (λ _, iff.rfl)),
-      exact m }
-  end
-end
+/-- A `fintype` structure on `insert a s` when inserting a new element. -/
```

Comment:
Can we reference `fintype_insert` in this docstring (and the next), and also vice-versa?

Reply:
I'll do that in a followup PR